rpcs3/Emu/Cell/Modules/cellImeJp.cpp

```diff
@@ -1,283 +1,1286 @@
+
+	if (!container_id || !hImeJpHandle)
+	{
+		return CELL_IMEJP_ERROR_PARAM;
+	}
+
+	if (idm::check<ime_jp_handle>(container_id))
+	{
+		return CELL_IMEJP_ERROR_ALREADY_OPEN;
+	}
+
+	const auto manager = g_fxo->get<ime_jp_manager>();
+	std::lock_guard lock(manager->mutex);
+
+	bool success = false;
+	auto context = manager->containers.begin();
+
+	std::tie(context, success) = manager->containers.try_emplace(container_id, ime_jp_container());
+
+	if (!success)
+	{
+		return CELL_IMEJP_ERROR_ALREADY_OPEN;
+	}
+
+	if (addDicPath && addDicPath->path[0])
+	{
+		cellImeJp.warning("cellImeJpOpen dictionary path = %s", addDicPath->path);
+
+		context->second.dictionary_paths.push_back(std::string(addDicPath->path));
+	}
+
+	*hImeJpHandle = vm::cast(idm::make<ime_jp_handle>(container_id));
+
 	return CELL_OK;
 }
 
-s32 cellImeJpOpen2()
+error_code cellImeJpOpen2(sys_memory_container_t container_id, vm::ptr<CellImeJpHandle> hImeJpHandle, vm::cptr<CellImeJpAddDic> addDicPath)
 {
-	std::memset(s_ime_string, 0, sizeof(s_ime_string));
-	cellImeJp.error("cellImeJpOpen2()");
+	cellImeJp.todo("cellImeJpOpen2(container_id=*0x%x, hImeJpHandle=*0x%x, addDicPath=*0x%x)", container_id, hImeJpHandle, addDicPath);
+
+	if (!container_id || !hImeJpHandle)
+	{
+		return CELL_IMEJP_ERROR_PARAM;
+	}
+
+	if (idm::check<ime_jp_handle>(container_id))
+	{
+		return CELL_IMEJP_ERROR_ALREADY_OPEN;
+	}
+
+	const auto manager = g_fxo->get<ime_jp_manager>();
+	std::lock_guard lock(manager->mutex);
+
+	bool success = false;
+	auto context = manager->containers.begin();
+
+	std::tie(context, success) = manager->containers.try_emplace(container_id, ime_jp_container());
+
+	if (!success)
+	{
+		return CELL_IMEJP_ERROR_ALREADY_OPEN;
+	}
+
+	if (addDicPath && addDicPath->path[0])
+	{
+		cellImeJp.warning("cellImeJpOpen2 dictionary path = %s", addDicPath->path);
+
+		context->second.dictionary_paths.push_back(std::string(addDicPath->path));
+	}
+
+	*hImeJpHandle = vm::cast(idm::make<ime_jp_handle>(container_id));
+
 	return CELL_OK;
 }
 
-s32 cellImeJpOpen3()
+error_code cellImeJpOpen3(sys_memory_container_t container_id, vm::ptr<CellImeJpHandle> hImeJpHandle, vm::cpptr<CellImeJpAddDic> addDicPath)
 {
-	std::memset(s_ime_string, 0, sizeof(s_ime_string));
-	cellImeJp.error("cellImeJpOpen3()");
+	cellImeJp.todo("cellImeJpOpen3(container_id=*0x%x, hImeJpHandle=*0x%x, addDicPath=*0x%x)", container_id, hImeJpHandle, addDicPath);
+
+	if (!container_id || !hImeJpHandle)
+	{
+		return CELL_IMEJP_ERROR_PARAM;
+	}
+
+	if (idm::check<ime_jp_handle>(container_id))
+	{
+		return CELL_IMEJP_ERROR_ALREADY_OPEN;
+	}
+
+	const auto manager = g_fxo->get<ime_jp_manager>();
+	std::lock_guard lock(manager->mutex);
+
+	bool success = false;
+	auto context = manager->containers.begin();
+
+	std::tie(context, success) = manager->containers.try_emplace(container_id, ime_jp_container());
```

Comment:
```suggestion
	const auto [context, success] = manager->containers.try_emplace(container_id, ime_jp_container());
```
you probably saw this in cellVoice, you can change the syntax there as well.